xwax.1
.TH XWAX "1"
.SH NAME
xwax \- Digital vinyl on Linux
.SH SYNOPSIS
.B xwax
[\fIoptions\fR]
.SH DESCRIPTION
.P
xwax is vinyl emulation software for Linux. It allows DJs and
turntablists to playback digital audio files (MP3, Ogg Vorbis, FLAC,
AAC and more), controlled using a normal pair of turntables via
timecoded vinyls.
.SH OPTIONS
.P
The ordering of options is important. Most options apply to
subsequent music libraries or decks, which can be given multiple times.
See the
.B EXAMPLES
below.
.TP
.B \-l, \-\-crate \fIpath\fR
Scan for tracks or playlist at the given path. Tracks are grouped into
a crate in the music library.
.TP
.B \-\-timecode \fIname\fR
Use the named timecode for subsequent decks. See
.B \-\-help
for a list of
valid timecodes. You will need the corresponding timecode signal on
vinyl to control playback.
.TP
.B \-\-33
Set the reference playback speed for subsequent decks to 33 and one
third revolutions per minute. This is the default.
.TP
.B \-\-45
Set the reference playback speed for subsequent decks to 45
revolutions per minute.
.TP
.B \-\-[no\-]protect
Protect against loading new tracks or resetting the start of the track
if the deck is playing. Applies to subsequent decks, and the default is
off.
.TP
.B \-\-phono
Adjust the noise thresholds of subsequent decks to tolerate a
cartridge-level signal connected to a line-level audio interface. This
is a 'software pre-amp'. Unless your audio path has low noise, this
will give worse results or may not work at all; a true phono
pre-amplifier is always preferred.
.TP
.B \-\-line
Set noise thresholds of subsequent decks to standard audio levels.
This reverses the effect of the
.B \-\-phono
option, and is the default.
.TP
.B \-\-import \fIpath\fR
Execute the given program to load tracks for playing. The program
outputs a stream of signed, little-endian, 16-bit, 2 channel audio on
standard output.
.TP
.B \-\-scan \fIpath\fR
Execute the given program to scan music libraries. Applies to
subsequent use of the
.B \-\-crate
flag.
.TP
.B \-\-dummy
Create a deck which is not connected to any audio device, used
for testing.
.TP
.B \-\-lock\-ram
Lock into RAM any memory required for real-time use.
This includes audio tracks held in memory which can be large.
Use
.B ulimit \-l
to raise the kernel's memory limit to allow this.
.TP
.B \-\-rtprio \fIn\fR
Change the real-time priority of the process. A priority of 0 gives
the process no priority, and is used for testing only.
.TP
.B \-\-geometry [\fIn\fRx\fIn\fR][+\fIn\fR+\fIn\fR][/\fIf\fR]
Change the geometry of the display in size, position and scale (zoom)
respectively.
The size and position is passed
to SDL, which may use it to set the display mode, or size of an X window.
See the
.B EXAMPLES.
.TP
.B \-\-no\-decor
Request to the window manager to create a 'frameless' window which
does not have the regular controls such as title bars and buttons.
This can be useful in conjunction with the
.B \-\-geometry
flag for dedicated xwax appliances.
.TP
.B \-h, \-\-help
Display the help message and default values.
.SH "ALSA DEVICE OPTIONS"
.P
The following options are available only when xwax is compiled with
ALSA support.
.TP
.B \-\-alsa \fIdevice\fR
Create a deck which uses the given ALSA device (eg. plughw:0).
.TP
.B \-\-rate \fIhz\fR
Set the sample rate for subsequent decks.
The default is to prefer 48000Hz, but use 44100Hz if that is
not available.
To maintain best performance, only sample rates implemented by
the hardware are available.
.TP
.B \-\-buffer \fIsamples\fR
Set the ALSA buffer size for subsequent decks.
Smaller is better for lower latencies and greater responsiveness.
Set too low and audible glitches will occur, with reports of underruns
to the console.
.SH "JACK DEVICE OPTIONS"
.P
The following options are available only when xwax is compiled with
JACK support.
.TP
.B \-\-jack \fIname\fR
Create a deck which connects to JACK and registers under the given
name.
.P
xwax does not set the sample rate for JACK devices; it uses the sample
rate given in the global JACK configuration.
.SH "OSS DEVICE OPTIONS"
.P
The following options are available only when xwax is compiled with
OSS support.
.TP
.B \-\-oss \fIpathname\fR
Create a deck which uses the given OSS device (eg. /dev/dsp).
.TP
.B \-\-rate \fIhz\fR
Set the sample rate for subsequent decks. The default for OSS devices
is 48000Hz.
.TP
.B \-\-oss\-buffers \fIn\fR
Set the number of buffers used by the OSS device. Applies to
subsequent decks.
.TP
.B \-\-oss\-fragment \fIn\fR
Set the size of a buffer used by the OSS device. The actual buffer is
2^n bytes. Applies to subsequent decks.
.SH HARDWARE CONTROLLER OPTIONS
.P
The following options are available only when xwax is compiled
with ALSA support.
.TP
.B \-\-dicer \fIdevice\fR
Use one or two Dicer controllers connected as the given ALSA device
(eg. hw:Dicer). See the section
.B NOVATION DICER CONTROLS
for more information.
.P
Adding a hardware controller results in control over subsequent decks,
up to the limit of the hardware.
.SH KEYBOARD CONTROLS
.P
The playback of each deck (direction, speed and position) is
controlled via the incoming timecode signal from the turntables.
The keyboard provides additional controls.
.P
"C-" and "S-" means a keypress is combined with
the 'Control' or 'Shift' key, respectively.
.P
Record selection controls:
.TP
cursor up, cursor down
Move highlighted record up/down by one.
.TP
page up, page down
Scroll the record listing up/down by one page.
.TP
left cursor, right cursor
Switch to the previous/next crate of records.
.TP
tab
Toggle between the current crate and the 'All records' crate.
.TP
C-tab
Toggle sort mode between: artist/track name, BPM and 'playlist'
order. Playlist order is the order in which records were returned
from the scanner.
.TP
C-S-tab
Re-scan the currently selected crate.
.P
To filter the current list of records type a portion of a record
name. Separate multiple searches with a space, and use backspace to
delete.
.P
Deck-specific controls:
.TS
l l l l.
Deck 0	Deck 1	Deck 2
F1	F5	F9	Load currently selected track to this deck
F2	F6	F10	Reset start of track to the current position
F3	F7	F11	Toggle timecode control on/off
C-F3	C-F7	C-F11	Cycle between available timecodes
.TE
.P
The "available timecodes" are those which have been the subject of any
.B \-\-timecode
flag on the command line.
.P
Audio display controls:
.TP
+, \-
Zoom in/out the close-up audio meters for all decks.
.SH NOVATION DICER CONTROLS
.P
The Novation Dicer provides hardware control of cue points. The controls
are:
.TP
cue mode: dice button (1-5)
Jump to the specified cue point, or set it if unset.
.TP
loop-roll mode: dicer button (1-5)
"Punch" to the specified cue point, or set it if unset. Returns playback
to normal when the button is released.
.TP
mode button + dice button (1-5)
Clear the specified cue point.
.P
The dice buttons are lit to show that the corresponding cue point is
set.
.SH EXAMPLES
.P
2-deck setup using one directory of music and OSS devices:
.sp
.RS
xwax \-\-crate ~/music \-\-oss /dev/dsp \-\-oss /dev/dsp1
.RE
.P
As above, but using ALSA devices:
.sp
.RS
xwax \-\-crate ~/music \-\-alsa hw:0 \-\-alsa hw:1
.RE
.P
2-deck setup using a different timecode on each deck:
.sp
.RS
xwax \-\-crate ~/music \-\-timecode serato_2a \-\-alsa hw:0 \-\-timecode mixvibes_v2 \-\-alsa hw:1
.RE
.P
As above, but with the second deck at 45 RPM:
.sp
.RS
xwax \-\-crate ~/music \-\-timecode serato_2a \-\-alsa hw:0 \-\-timecode mixvibes_v2 \-\-45 \-\-alsa hw:1
.RE
.P
Default to the same timecode, but allow switching at runtime:
.sp
.RS
xwax \-\-crate ~/music \-\-timecode serato_2a \-\-timecode mixvibes_v2 \-\-alsa hw:0 \-\-alsa hw:1
.RE
.P
3-deck setup with the third deck at a higher sample rate:
.sp
.RS
xwax \-\-crate ~/music \-\-rate 48000 \-\-alsa hw:0 \-\-alsa hw:1 \-\-rate 96000 \-\-alsa hw:2
.RE
.P
Using all three device types simultaneously, one deck on each:
.sp
.RS
xwax \-\-crate ~/music \-\-alsa hw:0 \-\-oss /dev/dsp1 \-j jack0
.RE
.P
Scan multiple music libraries:
.sp
.RS
xwax \-\-crate ~/music \-\-crate ~/sounds \-\-crate ~/mixes \-\-alsa hw:0
.RE
.P
Scan a second music library using a custom script:
.sp
.RS
xwax \-\-crate ~/music \-\-scan ./custom-scan \-\-crate ~/sounds \-\-alsa hw:0
.RE
.P
Control two decks with Dicer hardware:
.sp
.RS
xwax \-\-dicer hw:Dicer \-\-alsa hw:0 \-\-alsa hw:1
.RE
.P
Use a high resolution and enlarge the user interface:
.sp
.RS
xwax \-\-geometry 1920x1200/1.8 \-\-alsa hw:0
.RE
.SH HOMEPAGE
http://xwax.org/
.SH AUTHOR
Mark Hills <mark@xwax.org>


CHANGES
* Introduce long command line flags for readability
* Migrate from SDL version 1 to version 2
* Fix a bug which imported non-stereo files incorrectly

v1.9 (2023-05-12)
-----------------

* Replaced ALSA '-m' flag with a new '--buffer' in samples
* Revised ALSA buffer handling to improve latencies
* Renamed '-r' flag to '--rate'
* Restrict ALSA sample rates to those implemented in hardware
* Fixes for compatibility with musl C library
* Minor fixes

Acknowledgements: Daniel Schürmann

v1.8 (2021-08-18)
-----------------

* Change to the license to GPL version 3
* Compatibility with Pioneer DJ "RekordBox" vinyls
* Fix a bug where decimal numbers are parsed in an unexpected way

Acknowledgements: Hugh Frater, Kilian, Yves Adler

v1.7 (2018-01-19)
-----------------

* Fix a bug that prevented dithering from being used
* Honour the system locale; allow the full character set in track names
* When searching, match characters in other locales

Acknowledgements: Stefan Berg-Johansen

v1.6 (2016-08-13)
-----------------

* Added timecode creator (for the adventurous)
* Correct interpretation of ALSA's buffer size
* Command-line flag to remove all window decorations (--no-decor)
* A dummy deck for test cases
* Internal restructing in preparation for new features
* Build without warnings on modern compilers

Acknowledgements: Daniel James, Alessio Treglia

v1.5 (2014-02-09)
-----------------

* Scan the music library in the background on startup
* Function to re-scan a crate
* Performance improovement to audio resampler
* Additional font paths
* Minor bug fixes

v1.4 (2013-06-08)
-----------------

* Scalable user interface
* Simple control of timecode levels for 'software pre-amp'
* Bug fix: MIDI interface now works when only JACK audio is used
* Dicer flag renamed to '--dicer'
* Minor fixes

Acknowledgements: Daniel Holbach

v1.3 (2012-11-29)
-----------------

* Tempo (BPM) field, and sorting and changing of sort modes
* Allow initial X window size and position to be set by the user
* Makefile modifications for packagers
* Track loading errors are reported to the main interface

Acknowledgements: Engine, Lukas Fleischer, Daniel Holbach, Matej Laitl

v1.2 (2012-03-23)
-----------------

* Scan directories only for known file extensions
* Support Novation 'Dicer' controller
* Cue points, including 'punch' feature
* Extra protection against skips: optionally lock memory into RAM

Acknowledgements: Novation, Olivier Gauthier, Christoph Krapp, Mitchell Smith

v1.1 (2012-01-30)
-----------------

* Bug fix: incorrect display of track time remaining
* Instant loading of duplicate tracks
* Optionally protect decks during playback
* Improvements to music selector
* Increase speed building look-up tables

v1.0 (2011-08-01)
-----------------

* Changing of timecode at runtime
* Improved parsing of vinyl track numbers
* Bug fix: affecting multiple decks with different sample rates
* Optimise timecode error checking during scratching
* Require realtime priority; don't start without it
* Internal restructuring

Acknowledgements: Robert Flechtner, Daniel Holbach, Lukas Fleischer

v0.9 (2011-04-19)
-----------------

* Internal cleanups
* Filtering of duplicate entries in the record library
* Scanning of ordered playlists
* Improved response when scratching
* New parsing rules for filenames
* A single button toggles timecode control on/off

Acknowledgements: Daniel Holbach, Robert Vettel

v0.8 (2010-11-08)
-----------------

* 45 RPM control
* Conversion of non-44100Hz MP3 files
* Code cleanups and minor fixes

Acknowledgements: Ewan Colsell, Robert Flechtner, Daniel Holbach, Matej Laitl

v0.7 (2010-02-26)
-----------------

* Multiple crates in music selector (Yves Adler)
* Fix a potential buffer overflow (Matej Laitl)
* Higher quality resampler
* Improved pitch stability over long mixes
* Installation script and man page for distributions
* Moved from Bitstream Vera to DejaVu font
* Minor fixes

Acknowledgements: Yves Adler, Matej Laitl

v0.6 (2009-09-03)
-----------------

* Modular scanning of music library
* Improved parsing of pathnames
* Decreased memory use of timecode decoder
* Minor fixes

Acknowledgements: Yves Adler

v0.5 (2009-07-03)
-----------------

* Configurable sample rates
* Timecode support for MixVibes vinyls
* Rewritten timecode decoder with 4x resolution
* Rewritten timecode tracking with improved accuracy
* JACK audio device support
* Clearer display of current position in the track overview
* Minor fixes

Acknowledgements: Daniel Fasnacht, Yves Adler

v0.4 (2008-05-07)
-----------------

* Timecode improvements

v0.3 (2007-12-04)
-----------------

* ALSA device support
* Timecode optimisations
* Minor fixes

v0.2 (2007-06-12)
-----------------

* First open-source release


xwax.c
/*
 * Copyright (C) 2025 Mark Hills <mark@xwax.org>
 *
 * This file is part of "xwax".
 *
 * "xwax" is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License, version 3 as
 * published by the Free Software Foundation.
 *
 * "xwax" is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <https://www.gnu.org/licenses/>.
 *
 */

#include <assert.h>
#include <locale.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/mman.h> /* mlockall() */

#include "alsa.h"
#include "controller.h"
#include "device.h"
#include "dicer.h"
#include "dummy.h"
#include "interface.h"
#include "jack.h"
#include "library.h"
#include "oss.h"
#include "realtime.h"
#include "thread.h"
#include "rig.h"
#include "timecoder.h"
#include "track.h"
#include "xwax.h"

#define DEFAULT_OSS_BUFFERS 8
#define DEFAULT_OSS_FRAGMENT 7

#define DEFAULT_ALSA_BUFFER 240 /* samples */

#define DEFAULT_PRIORITY 80

#define DEFAULT_IMPORTER EXECDIR "/xwax-import"
#define DEFAULT_SCANNER EXECDIR "/xwax-scan"
#define DEFAULT_TIMECODE "serato_2a"

#define ARRAY_SIZE(x) (sizeof(x) / sizeof(*x))

char *banner = "xwax " VERSION \
    " (C) Copyright 2025 Mark Hills <mark@xwax.org>";

size_t ndeck;
struct deck deck[3];

static size_t nctl;
static struct controller ctl[2];

static struct rt rt;

static double speed;
static bool protect, phono;
static const char *importer;
static struct timecode_def *timecode;

static void usage(FILE *fd)
{
    fprintf(fd, "Usage: xwax [<options>]\n\n");

    fprintf(fd, "Program-wide options:\n"
      "  --lock-ram          Lock real-time memory into RAM\n"
      "  --rtprio <n>        Real-time priority (0 for no priority, default %d)\n"
      "  --geometry <s>      Set display geometry (see man page)\n"
      "  --no-decor          Request a window with no decorations\n"
      "  -h, --help          Display this message to stdout and exit\n\n",
      DEFAULT_PRIORITY);

    fprintf(fd, "Music library options:\n"
      "  -l, --crate <path>  Location to scan for audio tracks\n"
      "  --scan <program>    Library scanner (default '%s')\n\n",
      DEFAULT_SCANNER);

    fprintf(fd, "Deck options:\n"
      "  --timecode <name>   Timecode name\n"
      "  --33                Use timecode at 33.3RPM (default)\n"
      "  --45                Use timecode at 45RPM\n"
      "  --[no-]protect      Protect against certain operations while playing\n"
      "  --line              Line level signal (default)\n"
      "  --phono             Tolerate cartridge level signal ('software pre-amp')\n"
      "  --import <program>  Track importer (default '%s')\n"
      "  --dummy             Build a dummy deck with no audio device\n\n",
      DEFAULT_IMPORTER);

#ifdef WITH_OSS
    fprintf(fd, "OSS device options:\n"
      "  --oss <device>      Build a deck connected to OSS audio device\n"
      "  --rate <hz>         Sample rate (default 48000Hz)\n"
      "  --oss-buffers <n>   Number of buffers (default %d)\n"
      "  --oss-fragment <n>  Buffer size to request (2^n bytes, default %d)\n\n",
      DEFAULT_OSS_BUFFERS, DEFAULT_OSS_FRAGMENT);
#endif

#ifdef WITH_ALSA
    fprintf(fd, "ALSA device options:\n"
      "  --alsa <device>     Build a deck connected to ALSA audio device\n"
      "  --rate <hz>         Sample rate (default is automatic)\n"
      "  --buffer <n>        Buffer size (default %d samples)\n\n",
      DEFAULT_ALSA_BUFFER);
#endif

#ifdef WITH_JACK
    fprintf(fd, "JACK device options:\n"
      "  --jack <name>       Create a JACK deck with the given name\n\n");
#endif

#ifdef WITH_ALSA
    fprintf(fd, "MIDI control:\n"
      "  --dicer <device>    Novation Dicer\n\n");
#endif

    fprintf(fd,
      "The ordering of options is important. Options apply to subsequent\n"
      "music libraries or decks, which can be given multiple times. See the\n"
      "manual for details.\n\n"
      "Available timecodes (for use with -t):\n"
      "  serato_2a (default), serato_2b, serato_cd,\n"
      "  pioneer_a, pioneer_b,\n"
      "  traktor_a, traktor_b,\n"
      "  mixvibes_v2, mixvibes_7inch\n\n"
      "See the xwax(1) man page for full information and examples.\n");
}

static void deprecated(const char **arg, const char *old, const char *new)
{
    if (strcmp(*arg, old))
        return;

    fprintf(stderr, "Command line flag '%s' is deprecated; using '%s'\n", old, new);
    *arg = new;
}

static struct device* start_deck(const char *desc)
{
    fprintf(stderr, "Initialising deck %zd (%s)...\n", ndeck, desc);

    if (ndeck == ARRAY_SIZE(deck)) {
        fprintf(stderr, "Too many decks.\n");
        return NULL;
    }

    return &deck[ndeck].device;
}

static int commit_deck(void)
{
    int r;
    struct deck *d;
    size_t n;

    /* Fallback to a default timecode. Don't initialise this at the
     * front of the program to avoid buildling unnecessary LUTs */

    if (timecode == NULL) {
        timecode = timecoder_find_definition(DEFAULT_TIMECODE);
        assert(timecode != NULL);
    }

    d = &deck[ndeck];

    r = deck_init(d, &rt, timecode, importer, speed, phono, protect);
    if (r == -1)
        return -1;

    /* Connect this deck to available controllers */

    for (n = 0; n < nctl; n++)
        controller_add_deck(&ctl[n], d);

    ndeck++;

    return 0;
}

int main(int argc, const char *argv[])
{
    int rc = -1, n, priority;
    const char *scanner, *geo;
    char *endptr;
    bool use_mlock, decor;

    struct library library;

#if defined WITH_OSS || WITH_ALSA
    unsigned int rate;  /* or 0 for 'automatic' */
#endif

#ifdef WITH_OSS
    int oss_buffers, oss_fragment;
#endif

#ifdef WITH_ALSA
    unsigned int alsa_buffer;
#endif

    fprintf(stderr, "%s\n\n" NOTICE "\n\n", banner);

    if (setlocale(LC_ALL, "") == NULL) {
        fprintf(stderr, "Could not honour the local encoding\n");
        return -1;
    }

    /* Explicit formatting for numbers; parsing and printing.  Match
     * the user's expectations, and the documentation */

    if (setlocale(LC_NUMERIC, "POSIX") == NULL) {
        fprintf(stderr, "Could not set numeric encoding\n");
        return -1;
    }

    if (thread_global_init() == -1)
        return -1;
    if (library_global_init() == -1)
        return -1;

    if (rig_init() == -1)
        return -1;
    rt_init(&rt);
    library_init(&library);

    ndeck = 0;
    geo = "";
    decor = true;
    nctl = 0;
    priority = DEFAULT_PRIORITY;
    importer = DEFAULT_IMPORTER;
    scanner = DEFAULT_SCANNER;
    timecode = NULL;
    speed = 1.0;
    protect = false;
    phono = false;
    use_mlock = false;

#if defined WITH_OSS || WITH_ALSA
    rate = 0; /* automatic */
#endif

#ifdef WITH_ALSA
    alsa_buffer = DEFAULT_ALSA_BUFFER;
#endif

#ifdef WITH_OSS
    oss_fragment = DEFAULT_OSS_FRAGMENT;
    oss_buffers = DEFAULT_OSS_BUFFERS;
#endif

    /* Skip over command name */

    argv++;
    argc--;

    while (argc > 0) {
        deprecated(&argv[0], "-33", "--33");
        deprecated(&argv[0], "-45", "--45");
        deprecated(&argv[0], "-a", "--alsa");
        deprecated(&argv[0], "-c", "--protect");
        deprecated(&argv[0], "-d", "--oss");
        deprecated(&argv[0], "-g", "--geometry");
        deprecated(&argv[0], "-i", "--import");
        deprecated(&argv[0], "-j", "--jack");
        deprecated(&argv[0], "-k", "--lock-ram");
        deprecated(&argv[0], "-q", "--rtprio");
        deprecated(&argv[0], "-s", "--scan");
        deprecated(&argv[0], "-t", "--timecode");
        deprecated(&argv[0], "-u", "--no-protect");
#ifdef WITH_OSS
        deprecated(&argv[0], "-b", "--oss-buffers");
        deprecated(&argv[0], "-f", "--oss-fragment");
#endif

        if (!strcmp(argv[0], "-h") || !strcmp(argv[0], "--help")) {
            usage(stdout);
            return 0;

#ifdef WITH_OSS
        } else if (!strcmp(argv[0], "--oss-fragment")) {

            /* Set fragment size for subsequent devices */

            if (argc < 2) {
                fprintf(stderr, "-f requires an integer argument.\n");
                return -1;
            }

            oss_fragment = strtol(argv[1], &endptr, 10);
            if (*endptr != '\0') {
                fprintf(stderr, "-f requires an integer argument.\n");
                return -1;
            }

            /* Fragment sizes greater than the default aren't useful
             * as they are dependent on DEVICE_FRAME */

            if (oss_fragment < DEFAULT_OSS_FRAGMENT) {
                fprintf(stderr, "Fragment size must be %d or more; aborting.\n",
                        DEFAULT_OSS_FRAGMENT);
                return -1;
            }

            argv += 2;
            argc -= 2;

        } else if (!strcmp(argv[0], "--oss-buffers")) {

            /* Set number of buffers for subsequent devices */

            if (argc < 2) {
                fprintf(stderr, "%s requires an integer argument.\n", argv[0]);
                return -1;
            }

            oss_buffers = strtol(argv[1], &endptr, 10);
            if (*endptr != '\0') {
                fprintf(stderr, "%s requires an integer argument.\n", argv[0]);
                return -1;
            }

            argv += 2;
            argc -= 2;
#endif

#if defined WITH_OSS || WITH_ALSA
        } else if (!strcmp(argv[0], "--rate") || !strcmp(argv[0], "-r")) {

            if (!strcmp(argv[0], "-r"))
                fprintf(stderr, "-r will be removed in future, use --rate instead\n");

            /* Set sample rate for subsequence devices */

            if (argc < 2) {
                fprintf(stderr, "--rate requires an integer argument.\n");
                return -1;
            }

            rate = strtoul(argv[1], &endptr, 10);
            if (*endptr != '\0') {
                fprintf(stderr, "--rate requires an integer argument.\n");
                return -1;
            }

            if (rate < 8000) {
                fprintf(stderr, "--rate must be a positive integer, in Hz.\n");
                return -1;
            }

            argv += 2;
            argc -= 2;
#endif

#ifdef WITH_ALSA
        } else if (!strcmp(argv[0], "-m")) {
            fprintf(stderr, "-m is no longer available, check the man page for --buffer in samples\n");
            return -1;

        } else if (!strcmp(argv[0], "--buffer")) {

            /* Set size of ALSA buffer for subsequence devices */

            if (argc < 2) {
                fprintf(stderr, "--buffer requires an integer argument.\n");
                return -1;
            }

            alsa_buffer = strtoul(argv[1], &endptr, 10);
            if (*endptr != '\0') {
                fprintf(stderr, "--buffer requires an integer argument.\n");
                return -1;
            }

            argv += 2;
            argc -= 2;
#endif

        } else if (!strcmp(argv[0], "--oss") || !strcmp(argv[0], "--alsa") ||
                  !strcmp(argv[0], "--jack"))
        {
            int r;
            struct device *device;

            /* Create a deck */

            if (argc < 2) {
                fprintf(stderr, "-%c requires a device name as an argument.\n",
                        argv[0][1]);
                return -1;
            }

            device = start_deck(argv[1]);
            if (device == NULL)
                return -1;

            /* Work out which device type we are using, and initialise
             * an appropriate device. */

            switch(argv[0][2]) {

#ifdef WITH_OSS
            case 'o':
                r = oss_init(device, argv[1], rate ? rate : 48000,
                             oss_buffers, oss_fragment);
                break;
#endif
#ifdef WITH_ALSA
            case 'a':
                r = alsa_init(device, argv[1], rate, alsa_buffer);
                break;
#endif
#ifdef WITH_JACK
            case 'j':
                r = jack_init(device, argv[1]);
                break;
#endif
            default:
                fprintf(stderr, "Device '%s' is not supported by this "
                        "distribution of xwax.\n", argv[0]);
                return -1;
            }

            if (r == -1)
                return -1;

            commit_deck();

            argv += 2;
            argc -= 2;

        } else if (!strcmp(argv[0], "--dummy")) {

            struct device *v;

            v = start_deck("dummy");
            if (v == NULL)
                return -1;

            dummy_init(v);
            commit_deck();

            argv++;
            argc--;

        } else if (!strcmp(argv[0], "--timecode")) {

            /* Set the timecode definition to use */

            if (argc < 2) {
                fprintf(stderr, "%s requires a name as an argument.\n", argv[0]);
                return -1;
            }

            timecode = timecoder_find_definition(argv[1]);
            if (timecode == NULL) {
                fprintf(stderr, "Timecode '%s' is not known.\n", argv[1]);
                return -1;
            }

            argv += 2;
            argc -= 2;

        } else if (!strcmp(argv[0], "--33")) {

            speed = 1.0;

            argv++;
            argc--;

        } else if (!strcmp(argv[0], "--45")) {

            speed = 1.35;

            argv++;
            argc--;

        } else if (!strcmp(argv[0], "--protect")) {

            protect = true;

            argv++;
            argc--;

        } else if (!strcmp(argv[0], "--no-protect")) {

            protect = false;

            argv++;
            argc--;

        } else if (!strcmp(argv[0], "--line")) {

            phono = false;

            argv++;
            argc--;

        } else if (!strcmp(argv[0], "--phono")) {

            phono = true;

            argv++;
            argc--;

        } else if (!strcmp(argv[0], "--lock-ram")) {

            use_mlock = true;
            track_use_mlock();

            argv++;
            argc--;

        } else if (!strcmp(argv[0], "--rtprio")) {

            if (argc < 2) {
                fprintf(stderr, "--rtprio requires an integer argument.\n");
                return -1;
            }

            priority = strtol(argv[1], &endptr, 10);
            if (*endptr != '\0') {
                fprintf(stderr, "--rtprio requires an integer argument.\n");
                return -1;
            }

            if (priority < 0) {
                fprintf(stderr, "Priority (%d) must be zero or positive.\n",
                        priority);
                return -1;
            }

            argv += 2;
            argc -= 2;

        } else if (!strcmp(argv[0], "--geometry")) {

            if (argc < 2) {
                fprintf(stderr, "%s requires an argument.\n", argv[0]);
                return -1;
            }

            geo = argv[1];

            argv += 2;
            argc -= 2;

        } else if (!strcmp(argv[0], "--no-decor")) {

            decor = false;

            argv++;
            argc--;

        } else if (!strcmp(argv[0], "--import")) {

            /* Importer script for subsequent decks */

            if (argc < 2) {
                fprintf(stderr, "%s requires an executable path "
                        "as an argument.\n", argv[0]);
                return -1;
            }

            importer = argv[1];

            argv += 2;
            argc -= 2;

        } else if (!strcmp(argv[0], "--scan")) {

            /* Scan script for subsequent libraries */

            if (argc < 2) {
                fprintf(stderr, "%s requires an executable path "
                        "as an argument.\n", argv[0]);
                return -1;
            }

            scanner = argv[1];

            argv += 2;
            argc -= 2;

        } else if (!strcmp(argv[0], "-l") || !strcmp(argv[0], "--crate")) {

            /* Load in a music library */

            if (argc < 2) {
                fprintf(stderr, "%s requires a pathname as an argument.\n", argv[0]);
                return -1;
            }

            if (library_import(&library, scanner, argv[1]) == -1)
                return -1;

            argv += 2;
            argc -= 2;

#ifdef WITH_ALSA
        } else if (!strcmp(argv[0], "--dicer")) {

            struct controller *c;

            if (nctl == sizeof ctl) {
                fprintf(stderr, "Too many controllers; aborting.\n");
                return -1;
            }

            c = &ctl[nctl];

            if (argc < 2) {
                fprintf(stderr, "Dicer requires an ALSA device name.\n");
                return -1;
            }

            if (dicer_init(c, &rt, argv[1]) == -1)
                return -1;

            nctl++;

            argv += 2;
            argc -= 2;
#endif

        } else {
            fprintf(stderr, "'%s' argument is unknown; try -h.\n", argv[0]);
            return -1;
        }
    }

#ifdef WITH_ALSA
    alsa_clear_config_cache();
#endif

    if (ndeck == 0) {
        fprintf(stderr, "You need to give at least one audio device to use "
                "as a deck; try -h.\n");
        return -1;
    }

    rc = EXIT_FAILURE; /* until clean exit */

    /* Order is important: launch realtime thread first, then mlock.
     * Don't mlock the interface, use sparingly for audio threads */

    if (rt_start(&rt, priority) == -1)
        return -1;

    if (use_mlock && mlockall(MCL_CURRENT) == -1) {
        perror("mlockall");
        goto out_rt;
    }

    if (interface_start(&library, geo, decor) == -1)
        goto out_rt;

    if (rig_main() == -1)
        goto out_interface;

    rc = EXIT_SUCCESS;
    fprintf(stderr, "Exiting cleanly...\n");

out_interface:
    interface_stop();
out_rt:
    rt_stop(&rt);

    for (n = 0; n < ndeck; n++)
        deck_clear(&deck[n]);

    for (n = 0; n < nctl; n++)
        controller_clear(&ctl[n]);

    timecoder_free_lookup();
    library_clear(&library);
    rt_clear(&rt);
    rig_clear();
    library_global_clear();
    thread_global_clear();

    if (rc == EXIT_SUCCESS)
        fprintf(stderr, "Done.\n");

    return rc;
}
